### ISSUE TITLE
raw tracepoint program type

### ISSUE THREAD

**Issue by brendangregg** [iovisor/bpftrace#341]
Raw tracepoint program types were added in Linux 4.17. At some point we should support them, perhaps with a probe type: rawtracepoint, alias "rt".

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=c4f6699dfcb8558d138fe838f741b2c10f416cf9

They are faster than normal tracepoint programs, since they skip argument processing and give raw access to the arguments.

**Comment by brendangregg** [comment]
This came up today; we have a use case where we want the pre-translated tracepoint arguments.

**Comment by qais-yousef** [comment]
It'd be nice to have this feature. Some tracepoints are only accessible as raw tracepoints [1]

[1] https://lore.kernel.org/lkml/20210116182133.2286884-1-qais.yousef@arm.com/

**Comment by gghh** [comment]
Hello, I write here to be counted: adding access to raw tracepoints from bpftrace would be very desirable for my usecase.

@brendangregg above mentions the issue of performance (accessing raw tracepoints is faster), but as @qais-yousef says, tracepoints for the scheduler metric "PELT"[1] are available only in raw form.

Part of my work is on improving the frequency scaling governor "schedutil", which was born for ARM but is getting momentum on x86 too. Schedutil feeds on data from PELT and currently the only way to debug PELT is to load an external module to make the tracepoints visible or use BCC (with the addition of a recent kernel patch by Yousef for 5.12).

Brendan's latest book popularized bpftrace and this tool is getting a lot of traction; in some cases it is way more convenient to use than building and loading a module or writing a BCC script. It surely is for asking users to run a quick one-liner and report the result.

[1] Per-entity load tracking https://lwn.net/Articles/531853/

**Comment by fbs** [comment]
From a users perspective, how do we use this? I see some options that could work, without extending the list of builtins.

1:
```
rawtracepoint:something {
   @ = args[0]
}
```

2:
```
rawtracepoint:something {
   @ = args->raw[0]
}
```

3:
```
rawtracepoint:something {
   @ = arg0
}
```

1 & 2 stay closer to the tp syntax, using the `args` builtin. 2 would be easier to implement but 1 might be nicer to have. 3 makes it more like a kprobe, not sure if there is a limit to the amount of args a tracepoint can have but we can probably support `arg133030` if we want to.

@brendangregg @gghh

**Comment by qais-yousef** [comment]
My 2 pennies. 1 looks nicer to me, but if kprobes use 3 then IMHO it would be less confusing to stick to that convention. rawtracepoints and kprobes are very similar in behavior.

I think it's already there but I never got a chance to learn how it works properly, bpftrace will automatically query BTF when casting the args to struct rq for example which is a private data structure, right?

**Comment by fbs** [comment]
I'm not sure if we can use BTF in this case to get the tracepoint argument types as I think the  type is 
```
struct bpf_raw_tracepoint_args {
	__u64 args[0];
};
```.


But it would be nice to have

**Comment by bobrik** [comment]
Should this be closed now that #2542 is merged? It was released in v0.18.0 (and it's missing from the release notes).

**Comment by viktormalik** [comment]
> Should this be closed now that #2542 is merged?

Closing.

> It was released in v0.18.0 (and it's missing from the release notes).

Nice catch, @fbs could we add it retrospectively?
